### ISSUE TITLE
Use MemberData for methods with optional parameters

### ISSUE THREAD

**Issue by Diermeier** [xunit/xunit#1985]
Hello,

using a [Theory] test with [MemberData] which reference a method works fine. But if that method has optional parameters then the full signature call is required in [MemberData]-Attribute.

The below code results in a non-testable situation:

```csharp
    public class MyTestClass
    {
        public static IEnumerable<object[]> MyTestCases(string name, int maxSzenarios = 3)
        {
            for (int i = 1; i <= maxSzenarios; i++)
            {
                yield return new object[] { name, i };
            }
        }

        [Theory]
        [MemberData(nameof(MyTestCases), "MyFirst")]
        [MemberData(nameof(MyTestCases), "MySecond")]
        public void MyTestMethod(string name, int szenario)
        {
            // Do anything ...
        }
    }
```

This works:

```csharp
    public class MyTestClass
    {
        public static IEnumerable<object[]> MyTestCases(string name, int maxSzenarios = 3)
        {
            for (int i = 1; i <= maxSzenarios; i++)
            {
                yield return new object[] { name, i };
            }
        }

        [Theory]
        [MemberData(nameof(MyTestCases), "MyFirst", 2)]
        [MemberData(nameof(MyTestCases), "MySecond", 3)]
        public void MyTestMethod(string name, int szenario)
        {
            // Do anything ...
        }
    }
```

Instead optional parameter I can use classic method overloading for MyTestCases. That would be OK for me. In that case a more specific error message would help.

**Comment by mateusz-dobrowolny** [comment]
Hi @bradwilson, @jnewkirk 

I took a look at this issue and I wrote PoC for the fix.
I'd like someone to take a look at this, and let me know if it is a right direction.
If yes - then I'll proceed with unit test.

I have it on my branch, but I have no rights to upload it. How is the best way to proceed? Should I copy - paste changes here?

Regards

**Comment by bradwilson** [comment]
@mateusz-dobrowolny You should be able to fork this project into your account, and then push your branch there for review.

**Comment by mateusz-dobrowolny** [comment]
@bradwilson  Thanks!
So I made a draft with description in the first comment.
https://github.com/xunit/xunit/pull/2756

**Comment by bradwilson** [comment]
Thanks! I'll take a look soon.

**Comment by bradwilson** [comment]
@mateusz-dobrowolny I took a look at your draft PR, and it does not fix the issue. It does raise an error message related to it, which I think is useful for v2 (where we don't plan to support this), so I will likely port it to the v2 code base. I will leave this issue open, because I _think_ we would like to actually support this in v3 (depending on the difficulty in doing so).

Thanks!

**Comment by mateusz-dobrowolny** [comment]
@bradwilson thanks for your comment.
I can eventually try to port it into V2, or you can do it - as you wish.

If you would like to have support for MemberData methods in V3, I'll be happy to work on that, and when I'll have some questions I'll ask them in the main issue's discussion

**Comment by bradwilson** [comment]
@mateusz-dobrowolny Thanks! I did the v2 port already and closed your POC PR. Feel free to ask questions here if/when you decide to try to tackle the v3 full support. 😄

**Comment by JamesTerwilliger** [comment]
Hi @bradwilson ! Thought I'd give this a try, came up with this:
https://github.com/xunit/xunit/pull/2802

**Comment by bradwilson** [comment]
Fixed by #2802.

Available in v3 `0.1.1-pre.305`
